### PR TITLE
[Native] Implement the macOS/Cocoa and Gtk File Dialogs

### PR DESCRIPTION
**Pull request by zbaylin**
**EDIT**: This is now fixed, and there is more information below.

This provides bindings to `NSOpenPanel`. This is in response to #444 by @Et7f3.

I'm _so_ close, but for some reason, I keep getting strange memory errors on any operations on the string(s) returned.

For example, the code
```reason
Console.log("WORKS:");
Console.log(x);
Console.log("CRASHES:");
Console.log("" ++ x);
```
results in
error

Hopefully some fresh eyes can help with this.

@Et7f3, any help would be greatly appreciated!

### CODE REVIEW

**Comment by Et7f3** [comment]
Can you select file extention ? like "*.txt". On windows we can. It may be useful to add function to generate the string if it haven't the same syntax.

**Comment by zbaylin** [comment]
yes you can -- I haven't pushed it yet but I will when I get back home.

File: src/Native/Dialog.re [hunk 1]
```diff
@@ -1,5 +1,7 @@
-//open Reglfw.Glfw;
-
 [@noalloc] external alertSupported: unit => bool = "revery_alertSupported";
 
-// [@noalloc] external alert: (NativeWindow.t, string) => unit = "revery_alert";
+[@noalloc]
+external openFiles:
+  (~startDirectory: string=?, ~fileTypes: list(string)=?, string) =>
```
Comment (Et7f3) [inline src/Native/Dialog.re]: Could the last arg be `option(string)` because it is optionnal on windows.

File: src/Native/dialog.cpp [hunk 2]
```diff
@@ -42,4 +55,43 @@ CAMLprim value revery_alert(value vWindow, value vMessage) {
...
+  }
+  char** fileList;
+#ifdef __APPLE__
+  fileList = revery_open_files_cocoa(startDirectory, NULL, title);
+#endif
+
+  if (fileList) {
+    CAMLlocal2(camlArr, mlData);
+
+    int len = sizeof(fileList) / sizeof(char*);
```
Comment (Et7f3) [inline src/Native/dialog.cpp]:
```suggestion
    int len = sizeof(fileList) / sizeof(*fileList);
``` less error prone. And change when you change `fileList` type

File: src/Native/dialog.cpp [hunk 3]
```diff
@@ -42,4 +55,43 @@ CAMLprim value revery_alert(value vWindow, value vMessage) {
...
+      int strl = strlen(fileList[i]);
+
+      mlData = caml_alloc_string(strl);
+
+      memcpy(String_val(mlData), fileList[i], strl);
+
+      Store_field(camlArr, i, mlData);
+    }
+
+    CAMLreturn(Val_some(camlArr));
```
Comment (Et7f3) [inline src/Native/dialog.cpp]: Nothing to free ? `fileList` for instance ?

Comment (zbaylin) [reply src/Native/dialog.cpp]: Fixed in latest commit.

File: src/Native/dialog.cpp (hunk 2 above)
Comment (zbaylin) [reply src/Native/dialog.cpp]: I changed how this worked, so it isn't needed anymore, but thank you!

File: src/Native/Dialog.re (hunk 1 above)
Comment (zbaylin) [reply src/Native/Dialog.re]: Oops! It's optional on macOS too. My bad -- fixed in latest commit.

**Comment by zbaylin** [comment]
I fixed the alloc error. It was a result of me mistakenly adding `[@noalloc]` before my external declaration. Now the dialog is fully functional. I will add more functionality to it (changing the "OK" button text, allowing opening directories, disallowing opening files, etc.)

File: src/Native/dialog.cpp [hunk 4]
```diff
@@ -13,33 +14,123 @@
...
+    int canChooseFiles = Bool_val(vCanChooseFiles);
+    int canChooseDirectories = Bool_val(vCanChooseDirectories);
+
+    if (vFileTypes != Val_none) {
+        CAMLlocal1(camlArr);
+        camlArr = Some_val(vFileTypes);
+        fileTypesSize = Wosize_val(camlArr);
+
+        // Allocate space for an array
+        fileTypes = (char **)malloc(sizeof(char *) * fileTypesSize);
```
Comment (Et7f3) [inline src/Native/dialog.cpp]:
```suggestion
        fileTypes = (char **)malloc(sizeof(*fileTypes) * fileTypesSize);
```
less error prone.

File: src/Native/dialog.cpp [hunk 5]
```diff
@@ -13,33 +14,123 @@
...
+
+        // Populate the array with the CAML array;
+        for (int i = 0; i < fileTypesSize; i++) {
```
Comment (Et7f3) [inline src/Native/dialog.cpp]: Reverse loop + no need for local str :eyes: .

File: src/Native/dialog.cpp [hunk 6]
```diff
@@ -13,33 +14,123 @@
...
+    fileList = revery_open_files_cocoa(startDirectory, fileTypes, fileTypesSize,
+                                       allowMultiple, canChooseFiles,
+                                       canChooseDirectories, buttonText, title);
+#endif
+
+    if (fileList) {
+        CAMLlocal1(camlArr);
+
+        int len = -1;
+        while (fileList[++len] != NULL) {
```
Comment (Et7f3) [inline src/Native/dialog.cpp]:
```suggestion
        while (fileList[len]++ != NULL) {
```

File: src/Native/dialog.cpp [hunk 7]
```diff
@@ -13,33 +14,123 @@
...
+#ifdef __APPLE__
+    fileList = revery_open_files_cocoa(startDirectory, fileTypes, fileTypesSize,
+                                       allowMultiple, canChooseFiles,
+                                       canChooseDirectories, buttonText, title);
+#endif
+
+    if (fileList) {
+        CAMLlocal1(camlArr);
+
+        int len = -1;
```
Comment (Et7f3) [inline src/Native/dialog.cpp]:
```suggestion
        int len = 0;
```

File: src/Native/dialog.cpp [hunk 8]
```diff
@@ -13,33 +14,123 @@
...
+        while (fileList[++len] != NULL) {
+        }
+        camlArr = caml_alloc(len, 0);
+
+        for (int i = 0; i < len; i++) {
```
Comment (Et7f3) [inline src/Native/dialog.cpp]: reverse loop ?

File: src/Native/dialog_cocoa.c [hunk 9]
```diff
@@ -14,4 +14,80 @@ void revery_alert_cocoa(void *pWin, const char *szMessage) {
...
+    }
+
+    if (buttonText)
+        [panel setPrompt:[NSString stringWithCString:buttonText
+                                            encoding:NSUTF8StringEncoding]];
+
+    if (nsFileTypes) [panel setAllowedFileTypes:nsFileTypes];
+
+    // Run the actual panel/modal
+    NSInteger result = [panel runModal];
```
Comment (Et7f3) [inline src/Native/dialog_cocoa.c]: panel leak ?

Comment (zbaylin) [reply src/Native/dialog_cocoa.c]: fixed in latest commit

File: src/Native/dialog.cpp (hunk 4 above)
Comment (zbaylin) [reply src/Native/dialog.cpp]: Resolved in latest commit

**Comment by zbaylin** [comment]
I also am now adding bindings to the Gtk/Linux file dialog.

File: src/Native/dialog_cocoa.c [hunk 10]
```diff
@@ -14,4 +14,82 @@ void revery_alert_cocoa(void *pWin, const char *szMessage) {
...
+    int allowMultiple, int canChooseFiles, int canChooseDirectories,
+    int showHidden, const char *buttonText, const char *title) {
+    /* Creates an empty NSArray of filetypes (NSString's)
+        If [fileTypes] is not null, copy the C-strings to NSString's to the
+       NSArray
+    */
+    NSArray *nsFileTypes = NULL;
+    if (fileTypes) {
+        NSMutableArray *tmpArr =
+            [[NSMutableArray alloc] initWithCapacity:fileTypesSize];
```
Comment (Et7f3) [inline src/Native/dialog_cocoa.c]: Where it is released ? `tmpArr`

Comment (zbaylin) [reply src/Native/dialog_cocoa.c]: Yep!

File: src/Native/dialog_cocoa.c [hunk 11]
```diff
@@ -4,7 +4,7 @@
 #import <Cocoa/Cocoa.h>
 
 void revery_alert_cocoa(void *pWin, const char *szMessage) {
-    NSWindow* pCocoaWin  = (NSWindow *)pWin;
+    NSWindow *pCocoaWin = (NSWindow *)pWin;
 
     NSView *view = [[NSView alloc] init];
     NSAlert *alert = [[NSAlert alloc] init];
```
Comment (Et7f3) [inline src/Native/dialog_cocoa.c]: Should it bee released ?

Comment (zbaylin) [reply src/Native/dialog_cocoa.c]: This is not my code -- but I can modify it if you want.

File: src/Native/dialog_cocoa.c [hunk 12]
```diff
@@ -14,4 +14,82 @@ void revery_alert_cocoa(void *pWin, const char *szMessage) {
...
+    // Run the actual panel/modal
+    NSInteger result = [panel runModal];
+
+    [panel release];
+    [nsFileTypes release];
+
+    // If a file(s) was selected...
+    if (result == NSModalResponseOK) {
+        // ...get the list of URLs
+        NSArray *urls = [panel URLs];
```
Comment (Et7f3) [inline src/Native/dialog_cocoa.c]: Is `urls` automatically released ?

File: src/Native/dialog_cocoa.c (hunk 11 above)
Comment (Et7f3) [reply src/Native/dialog_cocoa.c]: Yeah I know but if it is a leak. I appreciate if this is fixed.

Comment (zbaylin) [reply src/Native/dialog_cocoa.c]: I don't think it's a leak. If you don't use `CAMLparam` & `CAMLreturn`, the OCaml garbage collector operates on the values passed into the function.

Comment (Et7f3) [reply src/Native/dialog_cocoa.c]: It if fine with OCaml but with the Obj-C part
> The rule is quite simple: you only need to release an object if you obtained it by using new, alloc, retain or copy (NARC for short). That's it. As soon as you're done with a NARC object and you don't need it anymore, you must release it. If you obtained the object though any other means, then you must not release it.

https://stackoverflow.com/a/6937337/7227940
For me I understand alloc/release like malloc/free in C.

File: src/Native/dialog_cocoa.c (hunk 12 above)
Comment (Et7f3) [reply src/Native/dialog_cocoa.c]: Seem fixed in 838db54.

File: src/Native/dialog_cocoa.c (hunk 11 above)
Comment (zbaylin) [reply src/Native/dialog_cocoa.c]: Ohh I see what you mean now.

Comment (zbaylin) [reply src/Native/dialog_cocoa.c]: Resolved in latest commit

**Comment by bryphe** [comment]
Thank you @zbaylin and @Et7f3 ! This is awesome!

**Pull request merged**